src/lib.rs

```diff
@@ -16,116 +16,261 @@
+        files_to_generate: &[String],
+        out_dir: &str,
+    ) {
+        use std::io::Write;
+
+        let output_dir = std::path::Path::new(out_dir);
+        let results = grpcio_compiler::codegen::gen(desc, &files_to_generate);
+        for res in results {
+            let out_file = output_dir.join(&res.name);
+            let mut f = std::fs::File::create(&out_file).unwrap();
+            f.write_all(&res.content).unwrap();
+        }
+    }
+
+    #[cfg(all(feature = "protobuf-codec", not(feature = "grpcio-protobuf-codec")))]
+    pub fn generate_grpcio(_: &[protobuf::descriptor::FileDescriptorProto], _: &[String], _: &str) {
+    }
 }
 
-/// Use Prost to generate Rust files from proto files (`file_names`).
-///
-/// Uses `["proto", "include"]` as the include lists.
-///
-/// Returns a list of the package names of the protocols that were compiled.
-pub fn generate_prost_files<T: AsRef<str>>(file_names: &[T], out_dir: &str) -> Vec<String> {
-    let packages = grpcio_compiler::prost_codegen::compile_protos(
-        &file_names.iter().map(AsRef::as_ref).collect::<Vec<_>>(),
-        &["proto", "include"],
+#[cfg(feature = "protobuf-codec")]
+pub use protobuf_imps::*;
+
+/// Use prost to generate Rust files from proto files (`files`).
+#[cfg(all(feature = "prost-codec", not(feature = "grpcio-prost-codec")))]
+pub fn generate<T: AsRef<Path>>(includes: &[T], files: &[T], out_dir: &str) {
+    prost_build::Config::new()
+        .out_dir(out_dir)
+        .compile_protos(files, includes)
+        .unwrap();
+    let mod_names = module_names_for_dir(out_dir);
+    generate_wrappers(
+        &mod_names
+            .iter()
+            .map(|m| format!("{}/{}.rs", out_dir, m))
+            .collect::<Vec<_>>(),
         out_dir,
-    )
-    .unwrap();
+        GenOpt::MUT
+            | GenOpt::HAS
+            | GenOpt::TAKE
+            | GenOpt::CLEAR
+            | GenOpt::MESSAGE
+            | GenOpt::TRIVIAL_SET,
+    );
+}
+
+/// TODO: merge this with prost-codec.
+#[cfg(feature = "grpcio-prost-codec")]
+pub fn generate<T: AsRef<Path>>(includes: &[T], files: &[T], out_dir: &str) {
+    let packages =
+        grpcio_compiler::prost_codegen::compile_protos(files, includes, out_dir).unwrap();
     for package in &packages {
-        let mut file_name = PathBuf::new();
+        let mut file_name = std::path::PathBuf::new();
         file_name.push(out_dir);
         file_name.push(&format!("{}.rs", package));
         rustfmt(&file_name);
     }
-
-    packages
 }
 
-pub fn rustfmt(file_path: &Path) {
-    let output = Command::new("rustfmt")
-        .arg(file_path.to_str().unwrap())
-        .output();
-    if !output.map(|o| o.status.success()).unwrap_or(false) {
-        eprintln!("Rustfmt failed");
+#[cfg(feature = "prost-codec")]
+mod prost_imps {
+    use super::wrapper::GenOpt;
+    use std::fs;
+    use std::path::Path;
+    use std::process::Command;
+
+    pub fn rustfmt(file_path: &Path) {
```

Comment:
we don't need this function any more

Reply:
The function is just in case the generated code is put in source directory again. But I don't have strong opinion about this.